### PR TITLE
PLT-6561 Rewrite the UTxO indexer using the new indexer interface

### PR DESCRIPTION
**Pull request by berewt**
The new UTxO indexer takes advantage of all the features discussed in Lisbon:

- The indexing logic is split into different specific indexers (`Utxo`, `Spent`, `Datum`, `BlockInfo`) and a SQL aggregate (`UtxoQuery`) gathers all the database for querying.
- The indexers are SQL only, we use the `WithCatchup` mechanism to perform batch-query before we reach the tip.
- We use a coordinator hierarchy to limit processing duplication across indexers.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Targeting main unless this is a cherry-pick backport
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [x] Reviewer requested

### CODE REVIEW

**Comment by berewt** [comment]
@koslambrou Kayvan asked to review this one. I was thinking about doing a code walkthrough during our tech meeting as well. Maybe it would be good if you'd review it as well?

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs [hunk 1]
```diff
@@ -404,6 +404,7 @@ propUtxoQueryShouldHaveSameDatumAsDatumQuery = Hedgehog.property $ do
   pure ()
 
 {- |
+ - TODO: test it in the event extraction logic, not in the indexer
```
Comment (koslambrou) [inline marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:407]: Did you want to do this in a separate PR?

File: marconi-core/src/Marconi/Core/Experiment/Indexer/SQLiteAggregateQuery.hs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+instance HasDatabasePath SQLiteIndexer where
+  getDatabasePath = Lens.view SQLite.databasePath
+
+type IsSourceProvider m event indexer = (IsSync m event indexer, HasDatabasePath indexer)
+
+{- | A wrapper that provides a @SQLiteAggregateQuery@ access to
+one of the indexers it requires to answer its query
+-}
+data SQLiteSourceProvider m point
```
Comment (koslambrou) [inline marconi-core/src/Marconi/Core/Experiment/Indexer/SQLiteAggregateQuery.hs:49]: Add comment that it's used for allowing heterogeous list in SQLiteAggregateQuery

File: marconi-chain-index/app/Experimental.hs [hunk 3]
```diff
@@ -1,28 +1,29 @@
...
-    sequence
-      [ fmap snd $ Indexers.utxoWorker (Cli.optionsDbPath o </> "utxo.db") securityParam
-      , fmap snd $ Mint.mintBurnTokenEventWorker (Cli.optionsDbPath o </> "minttokenevent.db")
-      ]
-
-  Indexers.runIndexers
+  mindexers <-
+    runExceptT $
+      buildIndexers
+        (Core.CatchupConfig 5000 100)
```
Comment (koslambrou) [inline marconi-chain-index/app/Experimental.hs:20]: Use vars for those magic numbers?

Comment (berewt) [reply marconi-chain-index/app/Experimental.hs:20]: Ideally, I would add it as CLI options in a next PR, I can use var instead at the moment.

File: marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs (hunk 1 above)
Comment (berewt) [reply marconi-chain-index/test/Spec/Marconi/ChainIndex/Indexers/Utxo/UtxoIndex.hs:407]: No, it was my notes to check which test I wanted to replicate or not in the new indexers, I'll remove them.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Coordinator.hs [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+coordinatorWorker
+  :: (MonadIO m, Ord (Core.Point b), Core.HasGenesis (Core.Point b))
+  => Text
+  -> (WithDistance a -> IO (Maybe b))
+  -> [Core.Worker b (Core.Point b)]
+  -> m (MVar (Core.Coordinator b), Core.Worker (WithDistance a) (Core.Point b))
+coordinatorWorker name extract workers = liftIO $ do
+  txBodyCoordinator <- Core.mkCoordinator workers
+  Core.createWorker name extract txBodyCoordinator
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Coordinator.hs:25]: It's because the coordinator handles workers that takes a `TxBody` as an input. I can either add a comment or find a better name.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/UtxoQuery.hs [hunk 5]
```diff
@@ -0,0 +1,316 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+{- | An aggregate indexer that gathers information of 'Utxo', 'Spent', 'Datum' and 'BlockInfo'
+to provide the UTxO information requested by sidechain in a single request.
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/UtxoQuery.hs:8]: It is quite specific, but it's probably easier to rename it at the moment and decide later whether we want to move it or not.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Coordinator.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Coordinator.hs:25]: But there's no TxBody in the types.

[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Coordinator.hs:25]: yes, but the coordinator transforms its input into `TxBody` and it's the type consumed by the workers, I updated the name to make it clearer.

[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/Coordinator.hs:25]: Oh my bad, I didn't notice the file... it's a dummy copy/paste issue. Fixing it. Sorry.

**Review by koslambrou — APPROVED** [review]
